### ISSUE TITLE
Rollback called when outside of a transaction when using deferred constraints

### ISSUE THREAD

**Issue by grongor** [doctrine/dbal#3423]
### Bug Report

In PostgreSQL there are deferrable constraints - constraints that are **checked at the time of commit**, not immediately after issuing an SQL statement.

And when the commit fails because of such constraint (duplicate key in unique constraint, for example) then the transaction is rolled back and an exception is thrown inside PDO.

The problem is that both DBAL and ORM don't expect that behaviour and they try to rollback the transaction after catching that exception. That triggers another exception - trying to rollback a transaction while there is no transaction active.

So this bug creates several issues:
- we try to rollback a transaction when there is no transaction
- we lose information about what error happened in the first place
- the Connection object is unusable after this happens as the internal "transaction level counter" remains at 1 while it should be 0

@simPod already discovered the issue and mentioned it here https://github.com/doctrine/doctrine2/issues/7545 . The problem with his issue is that he didn't find the root cause/only mentioned "previous exception being discarded", which is partly correct but doesn't reflect the severity of the issue. It's not like we only need to know "what happened before that last exception" - the second exception shouldn't have been thrown in the first place.

Code is better then 1000 words, so here you go:
- Failing build; added tests "explaining" the issue: https://travis-ci.org/grongor/dbal/builds/475298852
- Green build; added fix for the failing tests: https://travis-ci.org/grongor/dbal/builds/475299477
- Pull request with both the tests and the fix: https://github.com/doctrine/dbal/pull/3424

|    Q        |   A
|------------ | ------
| BC Break    | no
| Version     | all up to this day, afaik

#### Current behaviour

You can't use PostgreSQLs deferred constraints as they will kill DBAL/ORM when the constraint is violated.

#### How to reproduce

Create a table with deferrable unique constraint. Try to insert some data to the table when in transaction with deferrable constraint set to defer. Both DBAL and ORM will throw made up errors and end up in unexpected state.

#### Expected behaviour

I would expect to get an unique constraint violation exception. I would also expect to be able to use the connection after the fact.

**Comment by rela589n** [comment]
Has the fix been released yet?

Currently using 3.9.3, and it looks like it doesn't yet have it

**Comment by simPod** [comment]
No, there has been no release since.

**Comment by rela589n** [comment]
@greg0ire , could you please tell if there's any particular date when this is planned for release?

**Comment by greg0ire** [comment]
Yes: there is no particular date planned for any release we do (except major releases)

**Comment by rela589n** [comment]
would it be possible to release this particular fix as part of 3.9.4 version in the near future?

**Comment by greg0ire** [comment]
Yes

**Comment by greg0ire** [comment]
@rela589n done

**Comment by rela589n** [comment]
Thank you!
